File: Suyeon/ch10_Deque_Priority_queue/641_Design_Circular_Deque.py
class ListNode:
    def __init__(self, val=None, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class MyCircularDeque:
    def __init__(self, k: int):
        """
        Initialize your data structure here. Set the size of the deque to be k.
        """
        self.head = self.tail = None
        self.max = k
        self.len = 0

    def insertFront(self, value: int) -> bool:
        """
        Adds an item at the front of Deque. Return true if the operation is successful.
        """
        if self.isFull():
            return False

        if self.isEmpty(): # 비었을 때
            self.head = self.tail = ListNode(val=value)

        else:  # 아닐 때
            self.head.left = ListNode(val=value, right = self.head) # 새로운 노드의 오른쪽은 헤드
            self.head = self.head.left  # 기존헤드 옮기기
        self.head.left = self.tail
        self.tail.right = self.head


        self.len += 1

        return True

    def insertLast(self, value: int) -> bool:
        """
        Adds an item at the rear of Deque. Return true if the operation is successful.
        """

        if self.isFull():
            return False

        if self.isEmpty():  # 비었을 때
            self.head = self.tail = ListNode(val=value)

        else:  # 아닐 때
            self.tail.right = ListNode(val=value, left=self.tail)  # 새로운 노드의 왼쪽은 테일
            self.tail = self.tail.right  # 기존테일 옮기기

        self.head.left = self.tail
        self.tail.right = self.head
        self.len += 1

        return True

    def deleteFront(self) -> bool:
        """
        Deletes an item from the front of Deque. Return true if the operation is successful.
        """
        if self.isEmpty():
            return False

        if self.head == self.tail:
            self.head = self.tail = None
        else:
            self.head = self.head.right
            self.head.left = self.tail
            self.tail.right = self.head

        self.len -= 1

        return True

    def deleteLast(self) -> bool:
        """
        Deletes an item from the rear of Deque. Return true if the operation is successful.
        """
        if self.isEmpty():
            return False

        if self.head == self.tail:
            self.head = self.tail = None
        else:
            self.tail = self.tail.left
            self.head.left = self.tail
            self.tail.right = self.head

        self.len -= 1

        return True

    def getFront(self) -> int:
        """
        Get the front item from the deque.
        """
        if self.isEmpty():
            return -1

        return self.head.val

    def getRear(self) -> int:
        """
        Get the last item from the deque.
        """
        if self.isEmpty():
            return -1

        return self.tail.val

    def isEmpty(self) -> bool:
        """
        Checks whether the circular deque is empty or not.
        """

        # return self.head is None and self.tail is None
        return self.len == 0

    def isFull(self) -> bool:
        """
        Checks whether the circular deque is full or not.
        """
        # p = self.head
        #
        # for _ in range(self.max):
        #     if p is None:
        #         return False
        #     p = p.right
        #
        # return True
        return self.len == self.max



# Your MyCircularDeque object will be instantiated and called as such:
# obj = MyCircularDeque(k)
# param_1 = obj.insertFront(value)
# param_2 = obj.insertLast(value)
# param_3 = obj.deleteFront()
# param_4 = obj.deleteLast()
# param_5 = obj.getFront()
# param_6 = obj.getRear()
# param_7 = obj.isEmpty()
# param_8 = obj.isFull()

File: Sujin/Ch10_deque_and_priority_queue/641_design_circular_deque.cpp
#include <iostream>
using namespace std;

struct Node{
    int val;
    Node* next;
    Node(int v, Node*n = nullptr){
        val = v;
        next = n;
    }
};

class MyCircularDeque {
public:
    Node* front = NULL;
    Node* rear = NULL;
    int count;
    int maxCount;
    /** Initialize your data structure here. Set the size of the deque to be k. */
    MyCircularDeque(int k) {
        count=0;
        maxCount=k;
    }
    
    /** Adds an item at the front of Deque. Return true if the operation is successful. */
    bool insertFront(int value) {
        if(count==maxCount) return false;
        
        Node* nd = new Node(value);
        if(count==0){
            front = nd;
            rear = nd;
        }
        else{
            nd->next = front;
            front = nd;
        }
        count++;
        return true;
    }
    
    /** Adds an item at the rear of Deque. Return true if the operation is successful. */
    bool insertLast(int value) {
        if(count==maxCount) return false;
        
        Node* nd = new Node(value);
        if(count==0){
            front = nd;
            rear = nd;
        }
        else{
            rear->next = nd;
            rear = nd;
        }
        count++;
        return true;
    }
    
    /** Deletes an item from the front of Deque. Return true if the operation is successful. */
    bool deleteFront() {
        if(count==0) return false;
        
        Node* delNode = front; // delNode : 삭제할 노드를 가리킴
        front = front->next;
        delete(delNode);
        count--;
        return true;
    }
    
    /** Deletes an item from the rear of Deque. Return true if the operation is successful. */
    bool deleteLast() {
        if(count==0) return false;
        
        Node* delNode = rear;
        
        Node* cur = front; // cur : 삭제할 노드의 이전노드를 찾기위한 포인터
        while(cur->next!=rear && count>2) cur=cur->next;
        rear = cur; // rear 자리 다시 찾아주기
        
        delete(delNode);
        count--;
        return true;
    }
    
    /** Get the front item from the deque. */
    int getFront() {
        if(count==0) return -1;
        return front->val;
    }
    
    /** Get the last item from the deque. */
    int getRear() {
        if(count==0) return -1;
        return rear->val;
    }
    
    /** Checks whether the circular deque is empty or not. */
    bool isEmpty() {
        return count==0;
    }
    
    /** Checks whether the circular deque is full or not. */
    bool isFull() {
        return count==maxCount;
    }
};